### PR TITLE
Create outgoing requests

### PR DESCRIPTION
**Pull request by KassFlute**
This implements the backend, the frontend and the tests for the creation of borrow requests (outgoing requests).

### CODE REVIEW

**Comment by KassFlute** [comment]
Limitations:
- Descriptions are currently useless because Loans in the db don't have a description field.
- Items images are currently not implemented for the app.

Note:
- Start and end dates can be selected with the calendar icon on the right.
- Item name, owner name and Location can't be changed because the borrowers don't decide these.

File: app/src/androidTest/java/com/android/partagix/DatabaseTests.kt [hunk 1]
```diff
@@ -390,6 +390,54 @@ class DatabaseTests {
...
+
+    val loan =
+        Loan(
+            "id",
+            "id_owner",
+            "id_loaner",
+            "id_item",
+            Date(0),
+            Date(0),
+            "r",
```
Comment (LozanoCyrielle) [inline app/src/androidTest/java/com/android/partagix/DatabaseTests.kt:428]: les reviews sont des nombres il faudrait plutot mettre "0.0" pour la cohérence des tests

File: app/src/main/java/com/android/partagix/model/BorrowViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import com.android.partagix.model.item.Item
+import com.android.partagix.model.loan.Loan
+import com.android.partagix.model.loan.LoanState
+import com.android.partagix.model.user.User
+import com.android.partagix.model.visibility.Visibility
+import java.sql.Date
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+
+val emptyLoan = Loan("", "", "", "", Date(0), Date(0), "", "", "", "", LoanState.PENDING)
```
Comment (LozanoCyrielle) [inline app/src/main/java/com/android/partagix/model/BorrowViewModel.kt:16]: there empty variables already exists, in the model/emptyConst/empty.kt

File: app/src/main/java/com/android/partagix/model/BorrowViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  private var _userUiState = MutableStateFlow(emptyUser)
+  val userUiState: StateFlow<User> = _userUiState
+
+  /**
+   * Set the item to borrow and start a new borrow request
+   *
+   * @param item the item to borrow
+   */
+  fun resetBorrow(item: Item, owner: User) {
```
Comment (LozanoCyrielle) [inline app/src/main/java/com/android/partagix/model/BorrowViewModel.kt:38]: I feel like the name of the function doesn't really match, maybe more startBorrow to setToBorrow

**Review by LozanoCyrielle — CHANGES_REQUESTED** [review]
Great Work! There are some small changes to make that I put in comments in your code.
Also when choosing a date with the date picker, if you click "ok" but have not selected a date the app craches. I don't really know where it comes from, but maybe you can disable the okay button if no date are selected ? (because if you don't select a date and click on "cancel", the app doesn't craches)
A nice little addition would be to put today's date in default display and not Jan 1, 1970.

**Comment by KassFlute** [comment]
> ... Also when choosing a date with the date picker, if you click "ok" but have not selected a date the app craches. I don't really know where it comes from, but maybe you can disable the okay button if no date are selected ? (because if you don't select a date and click on "cancel", the app doesn't craches) A nice little addition would be to put today's date in default display and not Jan 1, 1970.

Do you have the last version of my branch? Because I've made the date pickers have today's date as default selected value in [fff9b3f](https://github.com/EPFLSWENT2024G1/partagix/pull/239/commits/fff9b3f5033dd851a37fb161af31169e9f50d189).

File: app/src/main/java/com/android/partagix/model/BorrowViewModel.kt (hunk 3 above)
Comment (KassFlute) [reply app/src/main/java/com/android/partagix/model/BorrowViewModel.kt]: I had struggle naming this function. `startBorrow` also makes more sense for me.

File: app/src/main/java/com/android/partagix/model/BorrowViewModel.kt (hunk 2 above)
Comment (KassFlute) [reply app/src/main/java/com/android/partagix/model/BorrowViewModel.kt]: Nice I didn't know. I will now use them.

File: app/src/androidTest/java/com/android/partagix/DatabaseTests.kt (hunk 1 above)
Comment (KassFlute) [reply app/src/androidTest/java/com/android/partagix/DatabaseTests.kt]: It makes sense, it's changed.

**Comment by LozanoCyrielle** [comment]
> > ... Also when choosing a date with the date picker, if you click "ok" but have not selected a date the app craches. I don't really know where it comes from, but maybe you can disable the okay button if no date are selected ? (because if you don't select a date and click on "cancel", the app doesn't craches) A nice little addition would be to put today's date in default display and not Jan 1, 1970.
> 
> Do you have the last version of my branch? Because I've made the date pickers have today's date as default selected value in [fff9b3f](https://github.com/EPFLSWENT2024G1/partagix/pull/239/commits/fff9b3f5033dd851a37fb161af31169e9f50d189). It crashed because of null pointer when getting the date from the date picker. But with today's date as default selected value the problem si solved.

I had the last version, but now I have the actual date, I don't now why

**Review by LozanoCyrielle — APPROVED** [review]
Nice work !

**Pull request merged**